Tcat zephyr implementation and openthread upmerge

canisLupus1313 opened the pull request:
This PR introduces new version of openthread `4ed44bc`.

Introduces also implementation of TCAT feature in Zephyr platform.

Commit:
net: openthread: Openthread upmerge to `4ed44bc`

This commit bumps openthread commit to `4ed44bc`
and implements `CONFIG_OPENTHREAD_MULTIPAN_RCP` option.

Signed-off-by: Przemyslaw Bida <przemyslaw.bida@nordicsemi.no>

edmont commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,508 @@
+/*
+ *  Copyright (c) 2022, The OpenThread Authors.
```
```suggestion
 *  Copyright (c) 2023, The OpenThread Authors.
```

edmont reviewed (commented):
It seems an initial compliance cleanup is needed.

Commit:
net: openthread: Add openthread TCAT implementation.

This commit adds bbtc implementation.
New file in modules/openthread/platform/ble.c
New corresponding kconfig option `OPENTHREAD_BLE_TCAT`.

Co-authored-by: Piotr Jasiński <piotr.jasinski@nordicsemi.no>

Signed-off-by: Przemyslaw Bida <przemyslaw.bida@nordicsemi.no>
Signed-off-by: Piotr Jasiński <piotr.jasinski@nordicsemi.no>

maciejbaczmanski reviewed (commented):
I'd suggest to keep all logs regarding errors as `LOG_WRN` level

rlubos commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,462 @@
+/* Zephyr Kernel Objects */
+
+static void otPlatBleThread(void *, void *, void *);
+uint8_t otPlatBleMsgBuf[PLAT_BLE_MSG_DATA_MAX];
+
+K_SEM_DEFINE(otPlatBleInitSemaphor, 0, 1);
+K_SEM_DEFINE(otPlatBleEventSemaphor, 0, 10000);
+RING_BUF_DECLARE(otPlatBleRingBuf, PLAT_BLE_RING_BUF_SIZE);
+K_THREAD_DEFINE(otPlatBleTid, PLAT_BLE_THREAD_STACK_SIZE, otPlatBleThread, NULL, NULL, NULL, 5, 0,
+		PLAT_BLE_THREAD_DEALY);
```
All those variables could be `static`? And please use `snake_case` for Zephyr code whenever possible (I am aware there have to exceptions for platform functions, but it's not the case for those variables).

canisLupus1313 replied:
fixed.

rlubos commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,462 @@
+
+	uint8_t len;
+	int8_t rssi;
+	otBleRadioPacket myPacket;
```
Ditto, snake_case plase.

canisLupus1313 replied:
fixed.

canisLupus1313 commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,462 @@
+
+	k_sched_lock();
+
+	if (ring_buf_space_get(&otPlatBleRingBuf) >=
```
fixed.

canisLupus1313 commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,462 @@
+static void otPlatBleThread(void *, void *, void *);
+uint8_t otPlatBleMsgBuf[PLAT_BLE_MSG_DATA_MAX];
+
+K_SEM_DEFINE(otPlatBleInitSemaphor, 0, 1);
```
fixed.

canisLupus1313 commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,462 @@
+
+/* Zephyr BLE Message Queue and Thread */
+
+static bool otPlatBleQueueMsg(const uint8_t *aData, uint8_t aLen, int8_t aRssi)
```
Yes it's not yet implemented.

canisLupus1313 commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,462 @@
+	err = bt_enable(bt_ready);
+
+	if (err != 0 && err != -EALREADY) {
+		LOG_INF("BLE enable failed with error code %d", err);
```
fixed.

canisLupus1313 commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,462 @@
+static void bt_ready(int err)
+{
+	if (err) {
+		LOG_INF("BLE init failed with error code %d", err);
```
fixed.

canisLupus1313 commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,462 @@
+	int err = bt_le_adv_start(BT_LE_ADV_CONN, ad, ARRAY_SIZE(ad), sd, ARRAY_SIZE(sd));
+
+	if (err != 0 && err != -EALREADY) {
+		LOG_INF("Advertising failed to start (err %d)", err);
```
fixed.

canisLupus1313 commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,462 @@
+	otPlatBleConnection = bt_conn_ref(conn);
+
+	if (err) {
+		LOG_INF("Connection failed (err %u)", err);
```
fixed.

canisLupus1313 commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,462 @@
+	/* Check whether notifications are enabled or not */
+	if (bt_gatt_is_subscribed(otPlatBleConnection, attr, BT_GATT_CCC_NOTIFY)) {
+		if (bt_gatt_notify_cb(otPlatBleConnection, &params)) {
+			LOG_INF("Error, unable to send notification");
```
fixed.

canisLupus1313 commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,462 @@
+			return OT_ERROR_INVALID_ARGS;
+		}
+	} else {
+		LOG_INF("Warning, notification not enabled on the selected attribute");
```
fixed.

canisLupus1313 commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,462 @@
+		LOG_INF("Connection failed (err %u)", err);
+		return;
+	} else if (bt_conn_get_info(conn, &info)) {
+		LOG_INF("Could not parse connection info");
```
fixed.

canisLupus1313 commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,462 @@
+					       .len = aPacket->mLength,
+					       .func = on_sent};
+
+	LOG_INF("Send data, handle %d, len %d", attr->handle, aPacket->mLength);
```
fixed.

canisLupus1313 commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,462 @@
+/* This function is called whenever a Notification has been sent by the TX Characteristic */
+static void on_sent(struct bt_conn *conn, void *user_data)
+{
+	ARG_UNUSED(user_data);
```
fixed.

canisLupus1313 commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,462 @@
+	case BT_GATT_CCC_INDICATE:
+		break;
+
+	case 0:
```
fixed.

maciejbaczmanski commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,459 @@
+
+	while (1) {
+		k_sem_take(&ot_plat_ble_event_semaphore, K_FOREVER);
+		ring_buf_get(&ot_plat_ble_ring_buf, (uint8_t*)&len, sizeof(len));
```
```suggestion
		ring_buf_get(&ot_plat_ble_ring_buf, (uint16_t*)&len, sizeof(len));
```

maciejbaczmanski commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,459 @@
+	len = sizeof(aLen) + sizeof(aRssi) + ((aLen <= PLAT_BLE_MSG_DATA_MAX) ? aLen : 0);
+
+	if (ring_buf_space_get(&ot_plat_ble_ring_buf) >= len) {
+		ring_buf_put(&ot_plat_ble_ring_buf, (uint8_t*)&aLen, sizeof(aLen));
```
```suggestion
		ring_buf_put(&ot_plat_ble_ring_buf, (uint16_t *)&aLen, sizeof(aLen));
```

maciejbaczmanski replied:
nvm,type cast is ok, my bad

canisLupus1313 replied:
This is to silence the compiler warning since `ring_buf_get` takes `const uint8_t*` I have to cast `len`.

maciejbaczmanski reviewed (commented):
missed one more type cast

canisLupus1313 commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,462 @@
+#define DEVICE_NAME     CONFIG_BT_DEVICE_NAME
+#define DEVICE_NAME_LEN (sizeof(DEVICE_NAME) - 1)
+
+#define TOBLE_SERVICE_UUID 0xfffb
```
Added a comment that those values are taken from thread speck.

canisLupus1313 commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,459 @@
+static uint8_t ot_plat_ble_msg_buf[PLAT_BLE_MSG_DATA_MAX];
+
+static K_SEM_DEFINE(ot_plat_ble_init_semaphore, 0, 1);
+static K_SEM_DEFINE(ot_plat_ble_event_semaphore, 0, 10000);
```
fixed.

canisLupus1313 commented on modules/openthread/platform/ble.c:
```diff
@@ -0,0 +1,459 @@
+	}
+}
+
+/*Zephyr BLE service callbacks */
```
fixed.
